README.md
Mod-Ntraffic
===================

mod_ntraffic is an Apache module to record traffic statistics for each Apache virtual host or a grand total for the whole server.

The module collects the hits, the inbound bytes and the outbound bytes.

The data are made available through HTTP GET requests to a configurable handler which can return data in XML, Json or Plain Text formats.

The data are also stored eventually on the file system for persistence or easier collection if necessary.


Installation
------------

To build mod_ntraffic you need apxs2.
This is usually shipped with the sources of your Apache server. If you are using a package manager, this utility is usually shipped
with the development files and headers.
If you are a Debian addict, for example, this utility is usually shipped with the package *apache2-threaded-dev* or *apache2-prefork-dev*
If you are using a RPM based distribution, the package is usually called *apache-devel* or *httpd-devel*.

To build the module simply run:

    make

To install the module simply run (as root):

    make install

Eventually Use sudo if needed.

Configuration
-------------

The distribution contains a file called *ntraffic.conf* which contains all the following **global** configuration directives.

```ruby

<IfModule mod_ntraffic.c>
    NTrafficEnabled             On
    NTrafficRefreshInterval     120

    NTrafficDataDir             /var/spool/apache2/mod_ntraffic/

    NTrafficExcludeIP           127.0.0.1/8
    NTrafficExcludeIP           10.1.1.0/24
    NTrafficExcludeIP           192.168.1.2

    <Location /ntraffic-status>
        SetHandler ntraffic-status

        RewriteEngine Off
        Order allow,deny
        Allow from 127.0.0.1
        Allow from 10.0.0.0/16
    </Location>
</IfModule>

```

- **NTrafficEnabled** (On/Off)

    Disables or enables Ntraffic globally

- **NTrafficRefreshInterval** (seconds)

    Ntraffic stores the statistics on files (inside NTrafficDataDir). This values specify
    the minimum interval between two updates.
    Please note that the file is updated only after a request (no request implies no update)

- **NTrafficDataDir**

    This is where the statistic files are stored. If not specified, no files will be used.

- **NTrafficExcludeIP** (IP or Network sub-net)

    Ntraffic counts all requests from all hosts. This directive, which can be used several times,
    indicates which hosts or networks should not be logged.


The location part installs the ntraffic-status handler for every virtual hosts, enabling access restrictions.


Access the data
---------------

The ntraffic-status handler is the main way to access the traffic statistics.

If you have, for example, a virtual host called www.test.it then accessing the following URL:

    http://www.test.it/ntraffic-status

from an IP inside an allowed network (see the Allow directive in the example above) will show you the virtual host statistics in XML (default format):

```XML
<document type="ntraffic/xml">
    <ntraffic-data>
        <vhost name="www.test.it" hits="74" sent="17734" recvd="15019" />
    </ntraffic-data>
</document>

```

There are four URL parameters that you may pass to alter the output:

- **globals** allows you to see the collected statistics of all the visited virtual hosts. If no statistics have been collected for a single virtual host, it will not be included in the list.

- **plain** turns the output from XML to plain TEXT, tab separated

- **json** turns the output to JSON

- **flush** clears the virtual host statistics (or every statistic if 'globals' is used) to zero after displaying them.


The **plain** and the **json** options should not be used together.

As an example, the following URL will:

    http://www.test.it/ntraffic-status?globals&json&flush

display all the statistics collected for every virtual host in json format and will clear them after displaying.



Development
------------

- Source hosted at [GitHub](https://github.com/ctrix/apache-mod-ntraffic)
- Report issues, questions, feature requests on [GitHub Issues](https://github.com/ctrix/apache-mod-ntraffic/issues)


Authors
-------

[Massimo Cetra](http://www.ctrix.it/)

* * *


debian/changelog
libapache2-mod-ntraffic (0.6-1) unstable; urgency=low

  * New upstream version
  * Support for Apache 2.4

 -- Massimo Cetra <mcetra@gmail.com>  Fri, 18 Oct 2013 01:22:00 +0200

libapache2-mod-ntraffic (0.5-1) unstable; urgency=low

  * New upstream version
  * Add JSON to the output formats
  * Avoid data files for inexistent virtual hosts

 -- Massimo Cetra <mcetra@gmail.com>  Thu, 17 Oct 2013 01:40:00 +0200

libapache2-mod-ntraffic (0.4-1) unstable; urgency=low

  * More checks on the context of the configuration options
  * Don't update our stats when we call our handler
  * Add flush option to ntraffic-status handler

 -- Massimo Cetra <mcetra@gmail.com>  Thu, 15 Aug 2008 01:40:00 +0200

libapache2-mod-ntraffic (0.3-1) unstable; urgency=low

  * Added options "globals" and "plain" to ntraffic-status handler

 -- Massimo Cetra <mcetra@gmail.com>  Thu, 14 Aug 2008 18:00:17 +0200

libapache2-mod-ntraffic (0.2-1) unstable; urgency=low

  * Adds support for "/ntraffic-status" handler

 -- Massimo Cetra <mcetra@gmail.com>  Thu, 14 Aug 2008 15:43:17 +0200

libapache2-mod-ntraffic (0.1-1) unstable; urgency=low

  * Initial release

 -- Massimo Cetra <mcetra@gmail.com>  Thu, 14 Aug 2008 11:53:27 +0200


shm_pool.c
/*
 * mod_ntraffic - Traffic statistics collector for Apache
 *
 * Copyright (C) 2008-2013, Massimo Cetra <massimo.cetra at gmail.com>
 *
 * Version: MPL 2
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is mod_ntraffic
 *
 * The Initial Developer of the Original Code is
 * Massimo Cetra <massimo.cetra at gmail.com>
 *
 * Portions created by the Initial Developer are Copyright (C)
 * the Initial Developer. All Rights Reserved.
 *
 */

#include "mod_ntraffic.h"

struct shm_pool_data {
    void *data;
};

struct shm_pool {
    apr_shm_t *shm;
    size_t size;
    size_t used;
    void *mem;
};

/* ************************************************************************************** */

void shm_dump(shm_pool_t * spool) {
    ap_log_error(APLOG_MARK, APLOG_ERR, 0, NULL, MODULE_NAME " SHM size: %ld  used: %ld", (long) spool->size,
                 (long) spool->used);
}

void *shm_get_base(shm_pool_t * spool) {
    return spool->mem;
}

apr_status_t shm_pool_create(apr_pool_t * pool, size_t shm_size, shm_pool_t ** shpool) {
    apr_status_t status;
    apr_shm_t *shm;
    apr_size_t sz;
    apr_size_t retsize;
    shm_pool_t *shmp;
    //shm_pool_data_t *shmpd;

    /* Create shared memory block */
    sz = sizeof(struct shm_pool) + sizeof(struct shm_pool_data) + shm_size;

    status = apr_shm_create(&shm, sz, tmpnam(NULL), pool);
    if (status != APR_SUCCESS) {
        return status;
    }

    /* Check size of shared memory block */
    retsize = apr_shm_size_get(shm);

    if (retsize != sz) {
        return status;
    }

    void *base = apr_shm_baseaddr_get(shm);
    if (base == NULL) {
        return status;
    }
    memset(base, 0, retsize);

    // Here i have the BASE ADDRESS and SHM to keep around the processes

    /* Init shm block */
    shmp = base;
    shmp->mem = shmp + sizeof(shm_pool_t);
    shmp->size = shm_size - sizeof(shm_pool_t);
    shmp->used = 0;
    shmp->shm = shm;
    //shmpd = shmp->mem;

    *shpool = shmp;

    return APR_SUCCESS;
}

void *shm_pool_alloc(shm_pool_t * sm, apr_size_t size) {
    void *mem = NULL;
    shm_pool_data_t *data;

    data = sm->mem;

    if (!sm || !data || size <= 0) {
        return NULL;
    }

    if (sm->size < sm->used + size) {
        return NULL;
    } else {
        mem = sm->mem + sm->used;
        sm->used += size;
    }

    return mem;
}

apr_status_t shm_pool_destroy(shm_pool_t * shm_pool) {
    /* If there was a memory block already assigned.. destroy it */
    apr_status_t status;

    if (shm_pool->shm) {
        status = apr_shm_destroy(shm_pool->shm);
        if (status != APR_SUCCESS) {
            return status;
        }
    }
    return APR_SUCCESS;
}


addons/mrtg/apache2mrtg.pl
#!/usr/bin/perl

#
# Parses apache config dir looking for virtual hosts.
# Then creates (on STDOUT) a simple configuration template
# for MRTG
#
# Copyright 2005 by Massimo Cetra <mcetra@gmail.com>.  All rights reserved.
#

my $CONFDIR="/etc/apache";	# Directory where httpd.conf is located
my $SCRIPTDIR="/usr/local/bin"; # no trailing slash
my $SCRIPTNAME="mod_ntraffic.pl";

#*******************************************
#*******************************************  DON'T EDIT BELOW HERE
#*******************************************
use strict;
use Cwd;


my %seen=();
my $topdir = $CONFDIR;
my $currdir = cwd();

chdir $topdir;
&dodir($topdir,0);

sub dodir {
    my ($dir,$nlink) = @_;
    my ($dev,$ino,$mode,$subcount);
    my( $filename );
    my @filenames;

    
    #print "****************************************\n";
    #print "Checking in $dir\n";
    #print "****************************************\n";

    # At the top level, we need to find nlink ourselves.
    ($dev,$ino,$mode,$nlink) = stat('.') unless $nlink;

    # Get the list of files in the current directory.

    opendir(DIR,'.') || die "Can't open $dir";
	@filenames = readdir(DIR);
    closedir(DIR);

    if ($nlink == 2) {        # This dir has no subdirectories.

        for (@filenames) {
            next if $_ eq '.';
            next if $_ eq '..';
            # print "Thatlevel: $dir/$_\n";
            my $filename = "$_";
            if ( -f $filename && -T _ ) { 
		$filename = "$dir/$_";
		ParseFile($filename);
	    }

        } 
    } else {                    # This dir has subdirectories.
        $subcount = $nlink - 2;
        for (@filenames) {
            next if $_ eq '.';
            next if $_ eq '..';
            my $name = "$dir/$_";
            my $filename = "$name";
            if ( -f $filename && -T _ ) { 
		ParseFile($filename);
	    }

            next if $subcount == 0;    # Seen all the subdirs?
            # Get link count and check for directoriness.
            ($dev,$ino,$mode,$nlink) = lstat($_);
            next unless -d _;
            # It really is a directory, so do it recursively.
            chdir $_ || die "Can't cd to $name";
            &dodir($name,$nlink);
            chdir '..';
            --$subcount;
        }
    }
}

#-------------------------------------------------------
sub ParseFile() {
    my ($file)=@_;
    my $name;

    open (FILE,$file);
    while (<FILE>) {
	# Skip commented blocks
	next if m!\#.*</?virtualhost!i;
	if (m!<virtualhost\s!i .. m!</virtualhost>!i) {
		$name = $1 if m!ServerName\s+(\S+)!;
	}
	if (m!</virtualhost>!i) {
		MakeTarget($name);
	}
}


    close(FILE);

}

#-------------------------------------------------------
sub MakeTarget() {
    my ($host)=@_;

    if ($seen{$host}) {
	#print STDERR "DUPLICATE entry \"$host\"\n";
	return;
    }

    $seen{$host} = 1;

    print <<EOT;
Title[$host]: $host Data Traffic
Target[$host]: `${SCRIPTDIR}/${SCRIPTNAME} $host`
MaxBytes[$host]: 1250000
PageTop[$host]: <h2><a href="http://$host/">$host</a> Data Traffic</h2>

EOT

}

1;


addons/mrtg/mod_ntraffic.pl
#!/usr/bin/perl

#
# Interface for apache mod_traffic and mrtg.
#
# Copyright 2005 by Massimo Cetra <mcetra@gmail.com>.  All rights reserved.
#

use strict;

my $dirfiles="/var/spool/apache/mod_ntraffic/";

#*******************************************
#*******************************************  DON'T EDIT BELOW HERE
#*******************************************

my $host=@ARGV[0];

if ( $host ne "" ) {
    ParseVhostData($host);
} else {
    print "0\n0\nNOT FOUND\nNOT FOUND\n";
}

#********************************************

sub ParseVhostData() {
    my ($vhost)=@_;
    my $line;

    my $file="$dirfiles".$vhost.".data";

    if ( ! -f $file ) {
	print "0\n0\nNOT FOUND\nNOT FOUND\n";
	return;
    }

    open (FILE,$file);
    $line=<FILE>;
    close(FILE);

    my ($sent,$rec,$hits)=split(" ",$line);


    print "$sent\n";
    print "$rec\n";
    # Parse though getuptime and get data
    my $getuptime = `/usr/bin/uptime`;
    my ($getuptime,$foo)=split(',',$getuptime);
    chomp($getuptime);
    print "$getuptime\n";
    print "$vhost\n";

}

1;


mod_ntraffic.h
/*
 * mod_ntraffic - Traffic statistics collector for Apache
 *
 * Copyright (C) 2008-2013, Massimo Cetra <massimo.cetra at gmail.com>
 *
 * Version: MPL 2
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is mod_ntraffic
 *
 * The Initial Developer of the Original Code is
 * Massimo Cetra <massimo.cetra at gmail.com>
 *
 * Portions created by the Initial Developer are Copyright (C)
 * the Initial Developer. All Rights Reserved.
 *
 */

#include <assert.h>

#include "httpd.h"
#include "http_config.h"
#include "http_request.h"
#include "http_protocol.h"
#include "http_core.h"
#include "http_main.h"
#include "http_log.h"
#include "unixd.h"

#include "mod_core.h"

#include "ap_release.h"
#include "ap_mpm.h"
#include "scoreboard.h"

#include "apr_strings.h"
#include "apr_atomic.h"

/* Uncomment to be flooded with error messages (to debug, obviously */
//#define DEBUG             1

#define MODULE_NAME		"mod_ntraffic"
#define MODULE_VERSION 		"0.6"
#define MODULE_COMPONENT_STR 	"mod_ntraffic/0.5"
#define DEFAULT_DATA_DIR 	"/www/var/spool/apache2/mod_traffic"
#define DEFAULT_UPDATE_INTERVAL 60
#define MAX_VHOSTS		8192
#define TOTALS_NAME		"__TOTALS__"
#define NTRAFFIC_MAGIC_TYPE 	"application/x-httpd-status"

/****************************************************************************    
    MAIN STRUCTURES
 ***************************************************************************/

typedef struct ntraffic_config_s ntraffic_config_t;
typedef struct traffic_data_s traffic_data_t;

typedef struct shm_pool_data shm_pool_data_t;
typedef struct shm_pool shm_pool_t;

struct ntraffic_config_s {
    char *name;
    short int vh;
    short int enabled;
    apr_uint32_t update_interval;
    char *data_dir;
    apr_array_header_t *exclude;
};

struct traffic_data_s {
    char *name;

    apr_uint32_t hits;
    apr_uint32_t recvd;
    apr_uint32_t sent;
    apr_time_t last_updated;

    traffic_data_t *next;
};

void shm_dump(shm_pool_t * spool);
void *shm_get_base(shm_pool_t * spool);
apr_status_t shm_pool_create(apr_pool_t * p, size_t shm_size, shm_pool_t ** shpool);
void *shm_pool_alloc(shm_pool_t * sm, apr_size_t size);
apr_status_t shm_pool_destroy(shm_pool_t * shm_pool);
